### PR TITLE
Pull the correct collection plugin for the product

### PR DESCRIPTION
**Pull request by jessicamack**
##### SUMMARY

There exist several cases of inventory plugins with the same name but different implementations. Historically, AWX/Controller pivoted on attributes on the inventory source to figure out which ansible inventory to call. Now, instead of a single implementation we've split them into two implementations completely. This PR is to implement the change to pick which one we want and when.

Requires: https://github.com/ansible/awx-plugins/pull/61

##### ISSUE TYPE

 - New or Enhanced Feature

##### COMPONENT NAME

 - Collection

##### AWX VERSION

```

```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: awx/main/models/inventory.py [hunk 1]
```diff
@@ -1404,5 +1405,10 @@ def get_absolute_url(self, request=None):
         return reverse('api:inventory_script_detail', kwargs={'pk': self.pk}, request=request)
 
 
-for cls in PluginFileInjector.__subclasses__():
-    InventorySourceOptions.injectors[cls.__name__] = cls
+awx_entry_points = {ep.name: ep for ep in entry_points(group='awx_plugins.inventory')}
+supported_entry_points = {ep.name: ep for ep in entry_points(group='awx_plugins.inventory.supported')}
+entry_points = awx_entry_points if detect_server_product_name() == 'AWX' else {**awx_entry_points, **supported_entry_points}
```
Comment (webknjaz) [inline awx/main/models/inventory.py]: Perhaps, it's better to skip the lookup altogether when not needed?
```suggestion
awx_entry_points = {ep.name: ep for ep in entry_points(group='awx_plugins.inventory')}
supported_entry_points = dict() if detect_server_product_name() == 'AWX' else {ep.name: ep for ep in entry_points(group='awx_plugins.inventory.supported')}
entry_points = {**awx_entry_points, **supported_entry_points}
```

File: awx/main/models/inventory.py [hunk 2]
```diff
@@ -1404,5 +1405,10 @@ def get_absolute_url(self, request=None):
...
+
+for entry_point_name, entry_point in entry_points.items():
+    cls = entry_point.load()
+    InventorySourceOptions.injectors[entry_point_name] = cls
```
Comment (webknjaz) [inline awx/main/models/inventory.py]: It'd be best if this wasn't happening during the import time but called in some predictable hook point, early in django's lifecycle.

File: awx/main/models/inventory.py (hunk 1 above)
Comment (webknjaz) [inline awx/main/models/inventory.py]: How about
```suggestion
def _load_all_entry_points_for(entry_point_subsections: list[str], /) -> dict[str, EntryPoint]:
    return {
        ep.name: ep
        for entry_point_category in entry_point_subsections
        for ep in entry_points(group=f'awx_plugins.{entry_point_category}')
    }

is_awx = detect_server_product_name() == 'AWX'
extra_entry_point_groups = () if is_awx else ('inventory.supported', )
entry_points = _load_all_entry_points_for(['inventory', *extra_entry_point_groups])
```
?

File: awx/main/models/inventory.py [hunk 3]
```diff
@@ -9,6 +9,7 @@
 import copy
 import os.path
 from urllib.parse import urljoin
+from importlib.metadata import entry_points
```
Comment (webknjaz) [inline awx/main/models/inventory.py]:
```suggestion
from importlib.metadata import entry_points, EntryPoint
```

**Review by webknjaz — COMMENTED** [review]
May I suggest a little refactoring?

File: awx/main/models/inventory.py (hunk 1 above)
Comment (webknjaz) [reply awx/main/models/inventory.py]: Although, it probably makes sense to put `_load_all_entry_points_for()` into `awx_plugins.interfaces`, making it reusable.

cc @chrismeyersfsu

File: awx/main/models/inventory.py (hunk 2 above)
Comment (chrismeyersfsu) [reply awx/main/models/inventory.py]: Yep, [here](https://github.com/ansible/awx/blob/cb04ad8ef5b5423ecf376925e77f4d3247af1cd9/awx/main/apps.py#L73) is a better place

File: awx/main/models/inventory.py (hunk 1 above)
Comment (chrismeyersfsu) [reply awx/main/models/inventory.py]: I lean towards not putting it in `awx_plugins.interfaces` right now. AWX is the only thing we plan on consuming these entry points at this time.

File: awx/main/models/inventory.py [hunk 4]
```diff
@@ -1404,5 +1405,10 @@ def get_absolute_url(self, request=None):
...
 
-for cls in PluginFileInjector.__subclasses__():
-    InventorySourceOptions.injectors[cls.__name__] = cls
+def _load_all_entry_points_for(entry_point_subsections: list[str], /) -> dict[str, EntryPoint]:
+    return {ep.name: ep for entry_point_category in entry_point_subsections for ep in entry_points(group=f'awx_plugins.{entry_point_category}')}
+
+
+is_awx = detect_server_product_name() == 'AWX'
+extra_entry_point_groups = () if is_awx else ('inventory.supported',)
+inv_entry_points = _load_all_entry_points_for(['inventory', *extra_entry_point_groups])
```
Comment (chrismeyersfsu) [inline awx/main/models/inventory.py]: We want to get away from running code during module init/load as much as possible.

I'd say move this code to `awx/main/apps.py` `load_inventory_plugins`.

Comment (jessicamack) [reply awx/main/models/inventory.py]: okay, I wasn't sure how much we were moving out at first. I'll move this as well

**Review by djyasin — APPROVED** [review]

**Comment by chrismeyersfsu** [comment]
The schema change failure here is a good thing. Prior to this PR we were in a state where we surfaced both the supported and unsupported version of the inventory plugins.

```
         "source": {
           "enum": [
             "azure_rm",
+            "controller",
             "ec2",
             "gce",
-            "vmware",
+            "insights",
+            "openshift_virtualization",
             "openstack",
             "rhv",
-            "rhv_supported",
             "satellite6",
-            "satellite6_supported",
             "terraform",
-            "controller",
-            "controller_supported",
-            "insights",
-            "insights_supported",
-            "openshift_virtualization",
-            "openshift_virtualization_supported",
+            "vmware",
             "scm",
             "constructed"
           ],
@@ -3336,21 +3331,16 @@
         "source": {
           "enum": [
             "azure_rm",
+            "controller",
             "ec2",
             "gce",
-            "vmware",
+            "insights",
+            "openshift_virtualization",
             "openstack",
             "rhv",
-            "rhv_supported",
make: *** [Makefile:548: detect-schema-change] Error 1
             "satellite6",
-            "satellite6_supported",
             "terraform",
-            "controller",
-            "controller_supported",
-            "insights",
-            "insights_supported",
-            "openshift_virtualization",
-            "openshift_virtualization_supported",
+            "vmware",
             "scm",
             "constructed"
           ],
@@ -3773,21 +3763,16 @@
         "source": {
           "enum": [
             "azure_rm",
+            "controller",
             "ec2",
             "gce",
-            "vmware",
+            "insights",
+            "openshift_virtualization",
             "openstack",
             "rhv",
-            "rhv_supported",
             "satellite6",
-            "satellite6_supported",
             "terraform",
-            "controller",
-            "controller_supported",
-            "insights",
-            "insights_supported",
-            "openshift_virtualization",
-            "openshift_virtualization_supported",
+            "vmware",
             "scm",
             "constructed"
           ],
```

**Review by chrismeyersfsu — APPROVED** [review]

File: awx/main/utils/common.py [hunk 5]
```diff
@@ -1248,3 +1249,7 @@ def _new_func(*args, **kwargs):
 
 def unified_job_class_to_event_table_name(job_class):
     return f'main_{job_class().event_class.__name__.lower()}'
+
+
+def load_all_entry_points_for(entry_point_subsections: list[str], /) -> dict[str, EntryPoint]:
```
Comment (AlanCoding) [inline awx/main/utils/common.py:1221]: Can you explain what the `/` is doing here? Maybe it's a new python thing I'm unaware of, but maybe it's unintended.

Comment (jessicamack) [reply awx/main/utils/common.py:1221]: my understanding is that it's using this syntax https://docs.python.org/3.8/whatsnew/3.8.html#positional-only-parameters. it was suggested by @webknjaz here https://github.com/ansible/awx/pull/15658#discussion_r1860965452. I think it's to enforce standard behavior.

**Review by AlanCoding — APPROVED** [review]

**Comment by pb82** [comment]
@jessicamack let's get this merged :+1:

**Pull request merged**